Correct clustermesh identity sync kvstore backend usage (to actually use the remote)

raybejjani opened the pull request:
_This PR is easier to review commit-by-commit. There are many changes but they are mostly moving behaviour around, or passing in the backend variable instead of using the global._

Clustermesh was broken. Routing, ipcache and node sync worked but policy did not. This meant policy drops because policy maps never saw the remote identities, but ipcache did and the datapath would match correctly.

Clustermesh identity sync accidentally used the cilium-agent local etcd backend. This was because `kvstoreBackend`, the middle-layer between the identity allocators and etcd/consul, accidentally called the `pkg/kvstore` functions. These functions then used the global etcd client, which is the one talking to the cluster-local etcd.
Clustermesh identity sync also accidentally never started. This was because `CachingIdentityAllocator.WatchRemoteIdentities` ignored the etcd backend passed in and called the watch functions on the global identity allocator. This had the added effect that it the allocator added itself to itself as a remote cluster allocator, causing identity list to double print identities.

ipcache and node sync worked because they used the correct, remote, etcd backend. The wiring around allocators did the wrong thing. This has likely been the case since the abstractions around CRD identities were added.

Edit 2020-02-14
Something is broken, but I don't know what. When testing with our clustermesh tutorial (https://cilium.readthedocs.io/en/latest/gettingstarted/clustermesh/) and these steps (https://gist.github.com/raybejjani/1272bea3e9d2beb40bcf4d115768b785) I see identities propagate across clusters but the policymap never gets remote identities. I even see log lines like `level=debug msg="ToMapState: Allowed remote IDs" EndpointSelector="&LabelSelector{MatchLabels:map[string]string{k8s.io.cilium.k8s.policy.cluster: cluster2,k8s.io.kubernetes.pod.namespace: default,k8s.name: rebel-base,},MatchExpressions:[]LabelSelectorRequirement{},}" policyID="[]" subsys=policy trafficDirection=Egress` but it never uses the remote IDs :/


Edit 2020-02-17
I just don't know. I'm pretty sure the issue is something in the policy resolution and not with clustermesh directly. Someone else should try this out in case I just spent 4 days debugging a typo in a policy. 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10185)
<!-- Reviewable:end -->

raybejjani committed:
kvstore/allocator: Keep internal reference to backend

We previously only used single pkg/kvstore/allocator.kvstoreBackend
instance system-wide. This also used a single, global, etcd/consul
backend instance (in this case a backend to allocator.kvstoreBackend,
itself the kvstore backend for the identity allocator).
With clustermesh, cilium-agent needs to sync with remote etcd instances.
This means that multiple etcd clients can exist, and that the identity
allocator may use different ones. This is now possible by keeping
an internal reference instead of using the global one.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
clustermesh: Use remote kvstore backend

Clustermesh remote-cluster watching accidentally used the local caching
identity allocator in cilium-agent. This meant that while ipcache and
nodes were sycned correctly, identities were not. This was because the
watch was called on the main identity allocator, adding itself to itself
as a remote cluster. This resulted in two connections to the local etcd,
and double entries when listing identities.

A new allocator instance is created to use the remote cluster etcd
backend. This is then added to the main allocators remote clusters list.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
kvstore: Move Trace calls into backends

We need to avoid using the package level functions but need to keep the
tracing for debug. Moving the trace calls into the kvstore backends is
verbose but is the simplest way to do this.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
kvstore: Replace package calls with .Client.function

We need to use different backends at the same time. Currently we assume a
single connected backend instance by using the package level functions.
We now direclty call these functions on the default client instead,
allowing future changes to use non-global instances of the backends.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

jrajahalme reviewed (approved):
Good work!

tgraf commented on pkg/kvstore/consul.go:
```diff
@@ -451,30 +451,40 @@ func (c *consulClient) Status() (string, error) {
-func (c *consulClient) Delete(ctx context.Context, key string) error {
+func (c *consulClient) Delete(ctx context.Context, key string) (err error) {
+	defer func() { Trace("Delete", err, logrus.Fields{fieldKey: key}) }()
+	return c.Delete(ctx, key)
```
Should this be `c.delete()`?

raybejjani replied:
yes it should! good catch

tgraf commented on pkg/kvstore/consul.go:
```diff
@@ -483,12 +493,15 @@ func (c *consulClient) Delete(ctx context.Context, key string) error {
+
 	duration := spanstat.Start()
 	qo := &consulAPI.QueryOptions{}
 	pair, _, err := c.KV().Get(key, qo.WithContext(ctx))
```
`s/:=/=/`

raybejjani replied:
`pair` is a new variable, though. If you're thinking that err won't be set correctly, I think that's only an issue with ifs (my test https://play.golang.org/p/QQSQpvv2C2P )

tgraf commented on pkg/kvstore/consul.go:
```diff
@@ -601,13 +651,17 @@ func (c *consulClient) CreateOnly(ctx context.Context, key string, value []byte,
+		Trace("CreateIfExists", err, logrus.Fields{fieldKey: key, fieldValue: string(value), fieldCondition: condKey, fieldAttachLease: lease})
+	}()
+
 	l, err := LockPath(ctx, condKey)
```
You are converting other `:=` to `=` but not the ones in this function

raybejjani replied:
oops. The trace should have been added to the public function just below this. I've switched it out. This function no longer needs the named err return variable.

tgraf reviewed (changes requested):
Looks good in general. I'm not sure we need a full-blown allocator for remote identities long-term but it's definitely fine as a fix to keep this minimal.

aanm commented on pkg/identity/cache/allocator.go:
```diff
@@ -437,5 +437,16 @@ func (m *CachingIdentityAllocator) ReleaseSlice(ctx context.Context, owner Ident
+
+	remoteAlloc, err := allocator.NewAllocator(GlobalIdentity{}, remoteAllocatorBackend)
+	if err != nil {
+		log.WithError(err).Errorf("Unable to initialize remote Identity Allocator")
```
Shouldn't this return nil here?

aanm commented on pkg/identity/cache/allocator.go:
```diff
@@ -437,5 +437,16 @@ func (m *CachingIdentityAllocator) ReleaseSlice(ctx context.Context, owner Ident
+
+	remoteAllocatorBackend, err := kvstoreallocator.NewKVStoreBackend(m.identitiesPath, m.owner.GetNodeSuffix(), GlobalIdentity{}, backend)
+	if err != nil {
+		log.WithError(err).WithField(logfields.BackendName, backend).Error("Error setting up remote allocator backend")
```
Shouldn't this return nil here?

raybejjani replied:
Yes, but it turned out to be worse. The caller of this didn't handle the nil case either. I made this return an error that is handled in the controller.

raybejjani committed:
clustermesh: Handle no-first connection in status

When the initial connection is being made and it is slow, or fails, the
remoteCluster is not fully initialized. This causes a segfault in the
status function. This is now handled.

Signed-off-by: Ray Bejjani <ray@isovalent.com>
